torch/distributed/_tensor/experimental/local_map.py

```diff
@@ -0,0 +1,174 @@
+            `PlacementType` values 1:1 mapping to the flattened `output`. For
+            :class:`Tensor` output, the corresponding `PlacementType` will be its
+            placements (a `Tuple[Placement]` value). For non-:class:`Tensor` output,
+            the `PlacementType` will be `None`.
+        in_placements (Union[`PlacementType`, Tuple[`PlacementType`, ...]], optional):
+            the required placements of the input :class:`DTensor`s. If not specified,
+            the input :class:`DTensor` will not be redistributed before passing its local
+            tensor to `func`. Similarly to `out_placements`, `in_placements` should keep
+            a 1:1 mapping to the flattened input of `func`. If a redistribution is
+            required according to `in_placements` and `redistribute_inputs` is `False`,
+            an exception will be raised.
+        device_mesh (:class:`DeviceMesh`, optional):
+            the device mesh that all the :class:`DTensor`s are placed on. If not
+            specified, this will be inferred from the input :class:`DTensor`s' device
+            mesh. `local_map` requires every :class:`DTensor`s to be placed on the same
+            device mesh.
+        redistribute_inputs (bool, optional):
+            the bool value indicating whether to reshard the input :class:`DTensor`s when
+            their placements are different from the required input placements. If this
+            value is `False` and some :class:`DTensor` input has a different placement,
+            an exception will be raised. Default: `False`.
+
+    Returns:
+        A `Callable` that applies `func` to each local shard of the input :class:`DTensor`
+        and returns a :class:`DTensor` constructed from the return value of `func`.
+
+    Raises:
+        AssertionError: If the input :class:`DTensor`s are not placed on the same device
+        mesh, or if they are placed on a different device mesh than the `device_mesh`
+        argument passed in.
+
+        AssertionError: If the placements in `out_placements` for non-:class:`DTensor`
+        return value is not `None`.
+
+        ValueError: If `redistribute_inputs=False` but the input :class:`DTensor` needs
+        a redistribution according to `in_placements`.
+
+    Example:
+        >>> # xdoctest: +SKIP("distributed")
+        >>> def mm_allreduce_forward(device_mesh, W, X):
+        >>>     partial_sum_tensor = torch.mm(W, X)
+        >>>     reduced_tensor = funcol.all_reduce(partial_sum_tensor, "sum", device_mesh)
+        >>>     return reduced_tensor
+        >>>
+        >>> W = torch.randn(12, 8, requires_grad=False)
+        >>> X = torch.randn(8, 16, requires_grad=False)
+        >>> Y = torch.mm(W, X)
+        >>> row_wise = [Shard(0)]  # row-wise sharding placements on 1-d mesh
+        >>> col_wise = [Shard(1)]  # col-wise sharding placements on 1-d mesh
+        >>>
+        >>> # d_mm_allreduce_forward is the function wrapped with DTensor/Tensor convertion
```

Comment:
nit: we should call this `local_mm_allreduce_forward`

Reply:
my understanding of the naming is:
`local_map` turns a "local" function into a function on DTensor. The new function should be a "d" function. WDYT?

Reply:
hmmm isn't it the reverse? 

`local_map` turns a function to a function that escapes DTensor and directly operate on local_tensor?

Reply:
after offline sync, we decide the naming:
1. `local_map` means this function takes in a `func` that computes on `torch.Tensor` and maps this `func` to its `DTensor` inputs' `_local_tensor`.
2. The return value of `local_map` is a function which 1) performs redistribution on `DTensor` inputs when necessary according to `input_placements`; 2) calls `func` on each `DTensor._local_tensor`; 3) wrap the return value back to `DTensor` by `DTensor.from_local(local_tensor, device_mesh, out_placement)`. This returned function should have prefix `local_` to emphasize that it operates on `DTensor`'s local tensor.